### ISSUE TITLE
Resource usage spike when setting up publications results in feedback loop and crashes

### ISSUE THREAD

**Issue by AndrewMorsillo** [meteor/meteor#10346]
Node v8.12.0 (also reproduced on 8.11.4)
Meteor 1.8.0 (also reproduced on 1.7.0.4)
I believe this issue impacts all meteor and node versions.

tl;dr -- clients disconnecting/reconnecting under heavy load sets up a feedback loop that crashes servers

I have realized that this may be an inherent flaw in Meteor. If the docs are correct, then anytime a client is disconnected for any reason and then reconnects the server will *not* resume/reuse their old publications and observers. It destroys the old ones and sets everything up from scratch again.

This means that if a server gets loaded heavily enough that some clients disconnect then they will immediately try to reconnect and upon successful reconnection load the server up even more by restarting all of their publications. It seems like this high-load feedback loop is built into the core of meteor and grows in severity with the per-user complexity of the app.

I'm running into this problem. My app is complex and uses a bunch of publications per user. It can handle lots of users at once with no problem as long as they don't all connect at the same time. Each user connecting and setting up its publications uses a ton of resources. This causes a problem on deployments or server crashes as there is a big spike in load as everybody reconnects. The resource usage when setting up publications is so high that users get disconnected then reconnect resulting in the feedback loop I'm describing. The only way I have around this at the moment is to vastly over-provision capacity per user just to handle the initial resource spike upon first connection. If an instance survives the spike then it hums along happily using ~10-20% of its capacity. I am potentially over provisioning by a factor of 10 just to get around this!

I'm having some trouble coming up with a good solution. Here are some alternatives:

1. When a client gets disconnected force them a click a button on the UI to reconnect. This is not good UX but it would probably spread the load out enough that the server wouldn't go into this feedback loop.
2. Try to implement session reconnection in meteor. I have no idea how complex this would be, but I doubt it would be easy.
3. Try to increase timeouts (in sockjs? i'm not sure where) so that under heavy load users don't lose their connection, they just experience severe lag until the load is gone. This also isn't a great UX.
4. Over-provision my instances and waste resources just to handle spikes of connecting users. This is what I do now but it results in a sky high cost to run per user, which I need to reduce.
5. Refactor my app to use less publications in order to avoid the resource usage spike caused by initializing them. I'd rather not do this as it would be a big time investment. The app runs fine with a bunch of users connected as long as they don't all connect at once. Just the huge spike of setting up publications is causing the problem.

Other than figuring out proper session reconnection none of these seem like great options. Anybody run into this issue or have thoughts?

**Comment by evolross** [comment]
Any idea why your users are disconnecting? Even if new users are data-intensive upon connection, that should affect existing users' connections.

**Comment by AndrewMorsillo** [comment]
I had some code that hit an express route to check the version number of my app. The purpose of this was to Meteor.disconnect() users when a new version was found and have them click a button to reload and reconnect. The thinking there was to spread the load out when a new version is deployed, which worked find.

What didn't work well was the fact that my version checks were set to time out after 4 seconds, and when that happened I was disconnecting the user until they made a successful check. These operations easily timed out under heavy load, disconnecting the user, then subsequently succeeded reconnecting them and re-starting all of their expensive publications. When I wrote that I had assumed that "reconnect" meant "reconnect to your session" not "destroy your old session and build a new one". That code made the issue significantly worse.

Even without my code making the problem worse it still happened with the default meteor configuration. I think meteor disconnects a user if their ping-pong keepalive fails. The default timeout for the heartbeat is 15 seconds which apparently I was surpassing when all users connected at once. This put them into the loop I described.

I (tentatively) seem to have solved this issue by doing a couple of things:
* Set the heartbeat interval and timeout to 60 seconds by adding the following to the top of the entrypoint of my server
```
Meteor.server.options.heartbeatInverval = 60000
Meteor.server.options.heartbeatTimeout = 60000
```
* Set the timeout for my version check call to 60 seconds
* If the version check fails, I delete the user's nginx routing cookie so they go to a different instance on reconnect

I just ran my first test with this. I scaled down and forced ~70 users to immediately connect to a new instance. Previously any more than ~35 would have hung the instance. This time CPU pegged at 100% but only for ~2 minutes while ~1500 observers were created. So far this instance is running smoothly after the initial spike, average CPU usage is 15-20%, ram is stable at ~2gb. 

I think the core of the problem still stands. There is a ton of overhead when a user adds a new publication. If you have a complex app and a lot of users connect at the same time then you will exceed the default connection timeout. When this happens the user is disconnected and reconnected which adds even more load to the already overloaded server. This cycle repeats until the server crashes or just stops responding because the GC is eating all the CPU time trying to clean up all the starting/stopping publications.

**Comment by AndrewMorsillo** [comment]
1. The ideal behavior would be for meteor to hang on to a connection for some duration after a client disconnect before cleaning it up. If the client reconnects before this timeout then resume their connection. If the client doesn't reconnect then destroy their connection. This could be a short interval, say 60 seconds, and it would prevent this issue of a disconnect/reconnect feedback loop forming.
2. I already use RedisOplog =)
3. I don't have a lot of observer re-use due to the nature of my app. Most of the publications are specific to the logged in user so they can't be re-used across users. The few publications that are shared have very high re-use.

As far as I see it there are two core issues:
1. Meteor destroys connections as soon as a client disconnects. This means if there is a brief disconnection then the server has to do a ton of work to build a new connection from the ground up.
2. The resource usage to set up new publications seems pretty extreme. If it weren't 50x (estimating) more expensive to set up a publication than it is to run one this wouldn't be an issue. This huge resource usage from setting up publications makes it difficult/impossible to deal with spikes in user connections without massively over-provisioning your servers.

**Comment by evolross** [comment]
Just curious - what are you using to load test?  How are you emulating 50 or 100 simultaneous users?

**Comment by AndrewMorsillo** [comment]
Hmm so I'm not out of the woods yet. I had quite a strange crash yesterday. I had an instance running along happily for nearly 24 hours with up to 100 users connected to it and then all of a sudden it went haywire! CPU pegged at 100% and ram usage began to climb steadily until it hit max_old_space_size and then totally stopped responding. 

<img width="1476" alt="screen shot 2018-11-23 at 3 30 23 pm" src="https://user-images.githubusercontent.com/858965/48959698-1df99080-ef35-11e8-802a-c5fc21005d4c.png">

I had to kill the instance to get things going again.

After some more digging I think this may be an issue with redis-oplog. An update with {multi: true} updated thousands of documents and it seems that sent my server into a death spiral due to the way redis-oplog handles multi updates.

**Comment by AndrewMorsillo** [comment]
I discovered a problem with my reactive publication strategy that was causing the crashing. I've removed all reactive publication from my project and things are a lot more stable now. The core issue of it being possible for meteor to enter a positive feedback loop of disconnection and reconnection leading to overload still stands IMO.

**Comment by evolross** [comment]
Just so I understand it... because your pub-sub data request was so heavy per user, Meteor would let the heartbeat time out on existing users and thus cause the connection feedback loop.  So the root issue is your heavy data requests by users and how that affects Meteor (and likely Node.js) keeping the heartbeat alive for existing users?

The connection/disconnection methodology as far as data seems sound, no?

**Comment by AndrewMorsillo** [comment]
This issue would crop up on any meteor server that got loaded heavily enough for a heartbeat to fail. It would be ideal if meteor didn't immediately delete a user's connection when contact is lost with a user. That would prevent an overloaded server from getting into a feedback loop that loads it further and prevents it from recovering from the load spike.

**Comment by ThomasRossi** [comment]
Hello, I've experienced the same exact behaviour (on helpless users as well), spike of reconnection.

@AndrewMorsillo could you please elaborate on what was the original "reactive publication strategy"? I'm not using anything fancy in my application, but still it would be interesting to see so I can compare.

The other super important number I should try to estimate is: how many container instances of a certain type I should prepare for a spike of n users. Let say 20 connections opening up at the same time on a page which has one subscription to one publication (a simple one, let say something like this one:

Meteor.publish('publicationbyids', function(arrayOfIds){
    check(arrayOfIds, [String]);
    let query = {
      _id: {$in: arrayOfIds}
    }
    let projection = {
      limit: 20, sort: { checkedInAt: -1 } 
    }
    return SomeTable.find( query, projection );
  })

I can tell you a single double instance won't do it, what were your experiments? So we can collect some data.. in my use case I can say exactly when the spikes will be so I can prepare exactly.

**Comment by AndrewMorsillo** [comment]
I first used [peerbrary:reactive-publish](https://github.com/peerlibrary/meteor-reactive-publish). I then built a custom solution [detailed here](https://forums.meteor.com/t/reactive-joins-without-using-any-packages/46525) that also didn't work well (it appeared to leak memory).

As for figuring out how many connections you can handle I think it really depends on your app. Is it OK for your use case to spend extra time loading? How complex is your app? The easiest way I know to do it is to use Kadira or Meteor APM and observe the behavior of your app when an instance restarts or some other connection spike happens.

**Comment by MartiniHenry1988** [comment]
@AndrewMorsillo I assume Kadira is no more available to use.
